src/journal/journald-server.c

```diff
@@ -450,40 +494,34 @@ static ManagedJournalFile* find_journal(Server *s, uid_t uid) {
 
         assert(s);
 
-        /* A rotate that fails to create the new journal (ENOSPC) leaves the rotated journal as NULL.  Unless
-         * we revisit opening, even after space is made available we'll continue to return NULL indefinitely.
+        if (s->storage == STORAGE_NONE)
+                return NULL;
+
+        /* A rotate that fails to create the new journal (ENOSPC) leaves the rotated journal as
+         * NULL.  Unless we revisit opening, even after space is made available we'll continue to
+         * return NULL indefinitely.
          *
-         * system_journal_open() is a noop if the journals are already open, so we can just call it here to
-         * recover from failed rotates (or anything else that's left the journals as NULL).
+         * system_journal_open() is a noop if the journals are already open, so we can just call it
+         * here to recover from failed rotates (or anything else that's left the journals as NULL).
          *
          * Fixes https://github.com/systemd/systemd/issues/3968 */
         (void) system_journal_open(s, false, false);
 
-        /* We split up user logs only on /var, not on /run. If the runtime file is open, we write to it
-         * exclusively, in order to guarantee proper order as soon as we flush /run to /var and close the
-         * runtime file. */
-
-        if (s->runtime_journal)
-                return s->runtime_journal;
-
-        /* If we are not in persistent mode, then we need return NULL immediately rather than opening a
-         * persistent journal of any sort.
-         *
-         * Fixes https://github.com/systemd/systemd/issues/20390 */
-        if (!IN_SET(s->storage, STORAGE_AUTO, STORAGE_PERSISTENT))
-                return NULL;
-
         if (!uid_for_system_journal(uid)) {
+                JournalStorage *storage;
                 ManagedJournalFile *f;
 
-                r = find_user_journal(s, uid, &f);
-                if (r >= 0)
-                        return ASSERT_PTR(f);
+                /* For now, system and user journals use the same storage */
+                storage = server_current_storage(s);
 
-                log_warning_errno(r, "Failed to open user journal file, falling back to system journal: %m");
+                r = find_user_journal(s, uid, storage, &f);
+                if (r >= 0 && f)
+                        return ASSERT_PTR(f);
```

Comment:
you literally checked that f is non-NULL one line before the ASSERT_PTR(). No need to do this immediately again, drop the ASSERT_PTR()